### DISCUSSION TITLE
Split up Renovate components

### DISCUSSION THREAD

**Discussion by Chumper** [renovatebot/renovate#13325] [Suggest an Idea]
During the log4j vulnerability I received multiple question from inside the company if Renovate can help mitigating the issue.
My answer was always that Renovate **can** help to update the dependency, but they need to be onboarded (in our case they needed to install the internal GitHub app). So Renovate was limited to suggesting the update of the dependency if the repo was onbarded.

Now It would be great if we can split up Renovate itself (in a backward compatible way) to provide each feature as their own binary.
In that case someone can use to find all dependencies within a project, or only get the updates possible or just get a repo with up to date branches.

This also has the advantage that we can provide a very "simple" extension mechanism. If people want to step in between fetching the updates and updating the branches, they can simply hook between the two components.

I currently can think of the following task Renovate does that could be moved into their own components:

* download
  * will download the repo from multiple sources (github, gitlab, etc), takes care of cloning if needed and provides it on a file system
* extract
  * will extract the dependencies from the files (filematch) and provides the information on stdout (or option) as json for the next component
* fetch
  * will fetch the list of updates based on the list of dependencies given on stdin (or option) and provides a list of updates on stdout (or option) as json for the next component
* update
  * will create the branches and updates the dependencies based on the list of possible updates given on stdin (or option)
* push
  * will push the local changes to the remote repository

other stuff Renovate does that I haven't thought about yet:
* automerge
* rebase

In the end I could imagine something like this:
```
renovate-download | renovate-extract | renovate-fetch | renovate-update | renovate-push
# with hooks
renovate-download | renovate-extract | renovate-fetch | ./myscript.sh | renovate-update | renovate-push
```

This would be backward compatible. The environment options will be the same for all components.

**Comment by HonkingGoose** [comment]
> During the log4j vulnerability I received multiple question from inside the company if Renovate can help mitigating the issue. My answer was always that Renovate **can** help to update the dependency, but they need to be onboarded (in our case they needed to install the internal GitHub app). So Renovate was limited to suggesting the update of the dependency if the repo was onbarded.
> 
> Now It would be great if we can split up Renovate itself (in a backward compatible way) to provide each feature as their own binary.

I don't understand this at all, I must be missing something important?

Normally, to use Renovate you _must_ be onboarded, or run the Renovate app yourself (self-hosting).
If you do not do either of these steps, you cannot use Renovate, that seems logical enough.

Why would splitting up the Renovate program help users who do not onboard/self-host? What is the benefit here?

---

Aside from the potential benefits, I worry that allowing users to mix and match Renovate components is going to make support and documentation really tricky.

I struggle to understand the behavior of Renovate as-is, and I'm helping to write/correct the documentation of the program! So if I already don't fully understand things, why should we expect our normal users to understand "mix and matching Renovate components", and expect our users to do this correctly?

**Comment by Chumper** [comment]
> I don't understand this at all, I must be missing something important?

https://security.googleblog.com/2021/12/understanding-impact-of-apache-log4j.html

There was a vulnerability in log4j that forces many companies to immediately update their systems.
Due to the wide use of log4j nearly every application was impacted.

> Why would splitting up the Renovate program help users who do not onboard/self-host? What is the benefit here?

What most people were missing was the information if their application was affected and if so how it can be updated.
I was approached to see if Renovate can help in two ways:

* Detect applications that depend on log4j
* Update applications that are using log4j

The first one is currently not possible with Renovate, the second one is.
You are right, that for this use case you need to be onboarded to Renovate.

Now Renovate does internally what a lot of people are looking for without using Renovate, which would be
* detect dependencies of an application, regardless of the language used (Renovate does that already for supported languages)
* check if dependencies of the applications have updates available (again, Renovate does that already) without having Renovate enabled
* update their dependencies without using Renovate

The benefit would be that Renovate could provide parts of its internal programming (dependency extraction, dependency update checking) to other people that are not using Renovate. This would have no immediate effect for Renovate itself but just by providing these building blocks we can enable other developers to build upon them if needed.

In my case I could have gone ahead and run `renovate-extract` on all internal repositories to check who is using log4j and based on that finding we could contact the teams to update them. That would be one use case.

Another one would be to maintain a list of dependencies somewhere (we have a internal portal for that, where each team needs to post their dependencies).
Something like `renovate-extract` could help to maintain that and with `renovate-fetch` we would even be able to notify teams on updates (in addition to Renovate on the repository)

> Aside from the potential benefits, I worry that allowing users to mix and match Renovate components is going to make support and documentation really tricky.

Depends, I think splitting Renovate up would make the responsibilities a lot clearer and easier do maintain and document, because each responsibility is clearly separated in their own binary (or module or whatever you want to call it).

Also allowing the mix & match would be only interesting for power users (like myself) that maintain Renovate for the whole company.
I would then be able to do the use cases above and could even use the data for other systems as well.
The default will still be that you just call `renovate` and it will work as expected, just under the hood it will combine the separated binaries/modules.

When I worked with Renovate I noticed that we are always passing the `config` around, which contains every setting available in Renovate.
While this is incredibly flexible it does not help in making clear what each part of Renovate does with it exactly.
By splitting up the code I would expect us to make the boundaries for each module strict and clear so that the maintenance will not suffer.

Currently I always need to debug complex situations because just from the code alone I am not able to see what Renovate does in detail.
Ideally we should be able to reason about the behavior of Renovate by looking at the code and not by debugging it.
This may be just my feeling as I am not that deep into Renovate, but that is what I noticed while working on it.

> I struggle to understand the behavior of Renovate as-is, and I'm helping to write/correct the documentation of the program! So if I already don't fully understand things, why should we expect our normal users to understand "mix and matching Renovate components", and expect our users to do this correctly?

Yes, this should be a power user feature only.

**Comment by BrunnerLivio** [comment]
Why don't you self-host Renovate instead? It's quite simple using the [provided docker image](https://hub.docker.com/r/renovate/renovate). The only tricky thing will be that Renovate somehow has access to all the repositories you'd like to target.

You can for instance have a global config where you disable most things (e.g. managers, etc.) and just target "log4j". On top of that you can disable the [onboarding PR](https://docs.renovatebot.com/self-hosted-configuration/), since the self-hosted global config would already take care of that.

> When I worked with Renovate I noticed that we are always passing the config around, which contains every setting available in Renovate.

[Why don't you use config presets](https://docs.renovatebot.com/config-presets/
)? 

> Currently I always need to debug complex situations because just from the code alone I am not able to see what Renovate does in detail.

It's much easier to debug if you self-host Renovate, since you have access to the logs and can use `LOG_LEVEL=debug`

Checkout more here how to do that:
https://github.com/renovatebot/github-action

**Comment by rarkins** [comment]
I think there's many ways we can improve Renovate's modularity, including especially not passing around the entire `config`, as you pointed out. I have a hesitance though to "expose" Renovate's modular internals much more, because I think it will make it ultimately slow down our progress if suddenly our external "contract" (which would determine e.g. semantic versioning) is substantially increased. We've done lots of refactoring and technical debt improvements (e.g. switching datasources to be class-based) which would be much harder and slower if we had to think about the versioning and backwards compatibility of all these internal interfaces.

I would like to leverage Renovate's internals more for purposes like you mention, such as extracting a simple dependency list, so in some cases we might try to "commit" to such interfaces not changing in future as needs arise.
